internal/graph/check.go

```diff
@@ -805,64 +829,151 @@ func buildTupleKeyConditionFilter(ctx context.Context, reqCtx *structpb.Struct,
+	usersetDetails := buildUsersetDetails(typesys, "", tuple.GetType(req.GetTupleKey().GetUser()))
+	return c.checkMembership(ctx, req, iter, usersetDetails)
+}
+
+type usersets struct {
+	err            error
+	objectRelation string
+	objectIDs      storage.SortedSet
+}
 
-	reqUserType := tuple.GetType(req.GetTupleKey().GetUser())
+func (c *LocalChecker) checkMembership(ctx context.Context, req *ResolveCheckRequest, iter *storage.ConditionsFilteredTupleKeyIterator, usersetDetails usersetDetailsFunc) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "checkMembership")
+	defer span.End()
 
+	// since this is an unbuffered channel, producer will be blocked until consumer catches up
+	// TODO: when implementing set math operators, change to buffered. consider using the number of sets as the concurrency limit
+	usersetsChan := make(chan usersets)
+	// usersetsMap is a map of all ObjectRelations and its Ids. For example,
+	// [group:1#member, group:2#member, group:1#owner, group:3#owner] will be stored as
+	// [group#member][1, 2]
+	// [group#owner][1, 3]
 	usersetsMap := make(usersetsMapType)
 
-	for {
-		// NOTE: For the future, once we observe a new ObjectRelation in the usersetsMap that means that all objectIDs
-		// needed for the intersection lookup for that ObjectRelation are in memory. There is no need to build the full
-		// usersetMap by draining the full iterator, it can start processing in batches of ObjectRelation.
-		// Consider doing it when we have better concurrency tooling for "streaming" operations.
-		t, err := iter.Next(ctx)
-		if err != nil {
-			if errors.Is(err, storage.ErrIteratorDone) {
-				break
+	cancellableCtx, cancelFunc := context.WithCancel(ctx)
+	// sending to channel in batches up to a pre-configured value of usersets to subsequently checkMembership for.
+	pool := concurrency.NewPool(cancellableCtx, 1)
+	pool.Go(func(ctx context.Context) error {
+		defer close(usersetsChan)
+		for {
+			t, err := iter.Next(ctx)
+			if err != nil {
+				if errors.Is(err, storage.ErrIteratorDone) {
+					break
+				}
+				// cancelled doesn't need to flush nor send errors back to main routine
+				if errors.Is(err, context.Canceled) {
+					return nil
+				}
+
+				telemetry.TraceError(span, err)
+				usersetsChan <- usersets{err: err}
+				return nil
+			}
+
+			objectRel, objectID, err := usersetDetails(t)
+			if err != nil {
+				usersetsChan <- usersets{err: err}
+			}
+
+			if _, ok := usersetsMap[objectRel]; !ok {
+				if len(usersetsMap) >= 1 {
+					// flush previous results
+					for k, v := range usersetsMap {
+						select {
+						case <-ctx.Done():
+							return nil
+						case usersetsChan <- usersets{
+							objectRelation: k,
+							objectIDs:      v,
+						}:
+							delete(usersetsMap, k)
+						}
+					}
+				}
+				usersetsMap[objectRel] = storage.NewSortedSet()
+			}
+
+			usersetsMap[objectRel].Add(objectID)
+			if usersetsMap[objectRel].Size() > 1000 {
```

Comment:
1000 is a magical number.  Please const it or make it configurable.  (Making it configurable allows better unit testing as none of the code below are covered by unit test).

Reply:
> none of the code below are covered by unit test

good catch, yikes

Reply:
Added more tests around context cancellation